### PR TITLE
PR: Move debug functions from IPython plugin to debugger plugin

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by impact27** [comment]
@ccordoba12 this is ready for review

File: spyder/plugins/debugger/plugin.py [hunk 1]
```diff
@@ -192,6 +198,15 @@ def on_main_menu_teardown(self):
 
     # ---- Private API
     # ------------------------------------------------------------------------
+    def _load_pdb_file_in_editor(self, fname, lineno):
+        """Load file using processevents."""
+        editor = self.get_plugin(Plugins.Editor)
+        if editor is None:
+            return
+        # Prevent keyboard input from accidentally entering the
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:206]:
```suggestion

        # Prevent keyboard input from accidentally entering the
```

File: spyder/plugins/debugger/plugin.py [hunk 2]
```diff
@@ -272,24 +287,23 @@ def _update_codeeditor(self, codeeditor):
         ):
             return
         # Update debugging state
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:289]:
```suggestion

        # Update debugging state
```

File: spyder/plugins/debugger/plugin.py [hunk 3]
```diff
@@ -382,3 +396,25 @@ def clear_breakpoint(self, filename, lineno):
...
 
         codeeditor.breakpoints_manager.toogle_breakpoint(lineno)
+
+    def can_close_file(self, filename=None):
+        """
+        Check if a file can be closed taking into account debugging state.
+        """
+        if not self.get_conf('pdb_prevent_closing'):
+            return True
+        widget = self.get_widget()
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:406]:
```suggestion

        widget = self.get_widget()
```

File: spyder/plugins/debugger/plugin.py [hunk 4]
```diff
@@ -382,3 +396,25 @@ def clear_breakpoint(self, filename, lineno):
...
+
+        debugging = widget.get_pdb_state()
+        if not debugging:
+            return True
+
+        pdb_fname, _ = widget.get_pdb_last_step()
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:412]:
```suggestion
        pdb_fname, __ = widget.get_pdb_last_step()
```

File: spyder/plugins/debugger/plugin.py [hunk 5]
```diff
@@ -382,3 +396,25 @@ def clear_breakpoint(self, filename, lineno):
...
+
+        if pdb_fname and filename:
+            if osp.normcase(pdb_fname) == osp.normcase(filename):
+                widget.print_debug_file_msg()
+                return False
+            return True
+        widget.print_debug_file_msg()
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:419]:
```suggestion

        widget.print_debug_file_msg()
```

File: spyder/plugins/debugger/utils/breakpointsmanager.py [hunk 6]
```diff
@@ -90,23 +90,21 @@ def __init__(self, editor):
...
-                return
+        if (
+            state
+            and filename
+            and self.filename
+            and osp.normcase(filename) == osp.normcase(self.filename)
+        ):
+            self.debugger_panel.start_clean()
+            self.debugger_panel.set_current_line_arrow(line_number)
+            return
```
Comment (ccordoba12) [inline spyder/plugins/debugger/utils/breakpointsmanager.py:107]:
```suggestion

            return
```

File: spyder/plugins/debugger/widgets/framesbrowser.py [hunk 7]
```diff
@@ -46,10 +46,46 @@ class FramesBrowser(QWidget, SpyderWidgetMixin):
...
+    Parameters
+    ----------
+    path: str
+        Path to file.
+    row: int
+        Cursor starting row position.
+    word: str
+        Word to select on given row.
+    """
+    sig_show_namespace = Signal(dict, object)
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/framesbrowser.py:63]:
```suggestion

    sig_show_namespace = Signal(dict, object)
```

File: spyder/plugins/debugger/widgets/framesbrowser.py [hunk 8]
```diff
@@ -46,10 +46,46 @@ class FramesBrowser(QWidget, SpyderWidgetMixin):
...
+    shellwidget: ShellWidget
+        The shellwidget the request originated from
+    """
     sig_update_actions_requested = Signal()
+    """Update the actions"""
     sig_hide_finder_requested = Signal()
+    """Hide the finder widget"""
+    sig_load_pdb_file = Signal(str, int)
+    """
+    Called when pdb reaches a new line
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/framesbrowser.py:80]:
```suggestion
    sig_update_actions_requested = Signal()
    """Update the widget actions."""

    sig_hide_finder_requested = Signal()
    """Hide the finder widget."""

    sig_load_pdb_file = Signal(str, int)
    """
    This signal is emitted when Pdb reaches a new line.
```

File: spyder/plugins/debugger/widgets/framesbrowser.py [hunk 9]
```diff
@@ -110,6 +157,10 @@ def setup(self):
         layout.addWidget(self.finder)
         self.setLayout(layout)
 
+    def _show_namespace(self, namespace):
+        """Request for the given namespace to be shown in variable explorer."""
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/framesbrowser.py:161]:
```suggestion
        """
        Request for the given namespace to be shown in the Variable Explorer.
        """
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 10]
```diff
@@ -85,8 +85,32 @@ class DebuggerWidget(ShellConnectMainWidget):
...
     sig_show_namespace = Signal(dict, object)
+    """
+    Show the namespace
+
+    Parameters
+    ----------
+    namespace: dict
+        A namespace view created by spyder_kernels
+    shellwidget: object
+        The shellwidget the resuest originated from
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:112]:
```suggestion
        The shellwidget the request originated from
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 11]
```diff
@@ -105,8 +129,31 @@ class DebuggerWidget(ShellConnectMainWidget):
     sig_clear_all_breakpoints = Signal()
     """Clear all breakpoints in all files."""
 
-    sig_pdb_state_changed = Signal(bool, dict)
-    """Pdb state changed"""
+    sig_pdb_state_changed = Signal(bool, str, int)
+    """
+    Called every time a pdb interaction happens
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:134]:
```suggestion
    This signal is emitted every time a Pdb interaction happens.
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 12]
```diff
@@ -105,8 +129,31 @@ class DebuggerWidget(ShellConnectMainWidget):
...
+
+    Parameters
+    ----------
+    pdb_state: bool
+        wether the debugger is waiting for input
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:139]:
```suggestion
        Whether the debugger is waiting for input
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 13]
```diff
@@ -105,8 +129,31 @@ class DebuggerWidget(ShellConnectMainWidget):
...
+    filename: str
+        The filename the debugger stepped in
+    line_number: int
+        The line number the debugger stepped in
+    """
+
+    sig_load_pdb_file = Signal(str, int)
+    """
+    Called when pdb reaches a new line
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:148]:
```suggestion
    This signal is emitted when Pdb reaches a new line.
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 14]
```diff
@@ -394,21 +439,29 @@ def create_new_widget(self, shellwidget):
         self.sig_breakpoints_saved.connect(widget.set_breakpoints)
 
         shellwidget.sig_pdb_state_changed.connect(self.sig_pdb_state_changed)
+
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:442]:
```suggestion
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27! My suggestions are purely stylistic, otherwise looks good to me.

**Comment by impact27** [comment]
@ccordoba12 I just changed `sig_pdb_state_changed` last-minute because it doesn't make sense to send the filename and line number when the pdb state changes

**Comment by impact27** [comment]
@ccordoba12 the Create Windows Installer tests are failing here too

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27!

**Pull request merged**